# Лабораторная работа 3. Проверка качества данных.

In [3]:
import pandas as pd
import numpy as np
import random
import matplotlib.pyplot as plt

from datetime import date, datetime, timedelta

## Задание 1: Проверка целостности данных.
Используйте набор данных о продажах и проверьте, заполнены ли все необходимые поля. Если обнаружены пропущенные значения, определите, какой процент они составляют от общего количества данных
 
Набор данных для использования: "Online Retail" на Kaggle. Этот набор данных представляет собой транзакции, произошедшие в период с декабря 2010 по октябре 2011 года и зарегистрированные розничной компанией из Великобритании.
Ссылка: https://www.kaggle.com/mashlyn/online-retail-ii-uci


In [6]:
df_retail = pd.read_csv(r'Datasets/online_retail_II.csv')

In [8]:
df_retail.head()
df_retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


Как можно заметить из общей информации о датасете, в столбцах Customer ID и Desctiprion общее число не пустых значений меньше, чем в остальных. 

Для достоверности посчитаем количество пустых полей в каждом из столбцов

In [10]:
for a in df_retail.columns:
    print(f'Пустых полей в колонке {a}: {df_retail[a].isna().sum()}')

Пустых полей в колонке Invoice: 0
Пустых полей в колонке StockCode: 0
Пустых полей в колонке Description: 4382
Пустых полей в колонке Quantity: 0
Пустых полей в колонке InvoiceDate: 0
Пустых полей в колонке Price: 0
Пустых полей в колонке Customer ID: 243007
Пустых полей в колонке Country: 0


Отсюда действительно видно, что в столбцах Description и Customer ID есть некотоорое число пустых полей. 

Для определения того, какой процент они составляют от общей массы, просто разделим количестов пустых полей на общее число данных.

In [13]:
for a in df_retail.columns:
    if any(df_retail[a].isna()):
        print(f'Пустые поля в {a} составляют {df_retail[a].isna().sum()/df_retail[a].count() * 100}%')

Пустые поля в Description составляют 0.4122338048653373%
Пустые поля в Customer ID составляют 29.478118889228544%


Таким образом, 0.41% описаний является пустым и ~29% айди также пусты.

## Задание 2. Проверка консистентности данных.

Рассмотрите набор данных о климате и проверьте, соблюдаются ли форматы дат во всех записях. 

Набор данных для использования: "Climate Change: Earth Surface Temperature Data" на Kaggle. 
Ссылка: https://www.kaggle.com/berkeleyearth/climate-change-earth-surface-temperature-data


In [134]:
df_temperatures = pd.read_csv(r'Datasets/GlobalTemperatures.csv')

In [136]:
df_temperatures.head()

dt  LandAverageTemperature  LandAverageTemperatureUncertainty  \
0  1750-01-01                   3.034                              3.574   
1  1750-02-01                   3.083                              3.702   
2  1750-03-01                   5.626                              3.076   
3  1750-04-01                   8.490                              2.451   
4  1750-05-01                  11.573                              2.072   

   LandMaxTemperature  LandMaxTemperatureUncertainty  LandMinTemperature  \
0                 NaN                            NaN                 NaN   
1                 NaN                            NaN                 NaN   
2                 NaN                            NaN                 NaN   
3                 NaN                            NaN                 NaN   
4                 NaN                            NaN                 NaN   

   LandMinTemperatureUncertainty  LandAndOceanAverageTemperature  \
0                            NaN                             NaN   
1                            NaN                             NaN   
2                            NaN                             NaN   
3                            NaN                             NaN   
4                            NaN                             NaN   

   LandAndOceanAverageTemperatureUncertainty  
0                                        NaN  
1                                        NaN  
2                                        NaN  
3                                        NaN  
4                                        NaN

In [147]:
df_temperatures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3192 entries, 0 to 3191
Data columns (total 9 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   dt                                         1800 non-null   object 
 1   LandAverageTemperature                     1788 non-null   float64
 2   LandAverageTemperatureUncertainty          1788 non-null   float64
 3   LandMaxTemperature                         600 non-null    float64
 4   LandMaxTemperatureUncertainty              600 non-null    float64
 5   LandMinTemperature                         600 non-null    float64
 6   LandMinTemperatureUncertainty              600 non-null    float64
 7   LandAndOceanAverageTemperature             600 non-null    float64
 8   LandAndOceanAverageTemperatureUncertainty  600 non-null    float64
dtypes: float64(8), object(1)
memory usage: 224.6+ KB


In [153]:
def format(df):
    list = []
    
    for a in df['dt']:
        if datetime.strptime(a, '%Y-%M-%d'):
            list.append('Correct format')
        else:
            list.append('Wrong format')
    df['format'] = list

format(df_temperatures)

df_temperatures['format'].value_counts()

TypeError: strptime() argument 1 must be str, not float

Все даты в предложенном датасете удовлетворяют формату ГГ-ММ-ДД.

## Задание 3. Проверка актуальности данных.

Анализируя набор данных о твитах, проверьте даты публикации твитов и определите, какие из них актуальны на текущий момент.

Набор данных для использования: "COVID19 Tweets" на Kaggle, включающий твиты о пандемии COVID-19.
Ссылка: https://www.kaggle.com/gpreda/covid19-twees


In [113]:
df_covid = pd.read_csv(r'Datasets/covid19_tweets.csv')

In [114]:
df_covid.head()

user_name         user_location  \
0           ᏉᎥ☻լꂅϮ            astroworld   
1    Tom Basile 🇺🇸          New York, NY   
2  Time4fisticuffs      Pewee Valley, KY   
3      ethel mertz  Stuck in the Middle    
4         DIPR-J&K     Jammu and Kashmir   

                                    user_description         user_created  \
0  wednesday addams as a disney princess keepin i...  2017-05-26 05:46:42   
1  Husband, Father, Columnist & Commentator. Auth...  2009-04-16 20:06:23   
2  #Christian #Catholic #Conservative #Reagan #Re...  2009-02-28 18:57:41   
3  #Browns #Indians #ClevelandProud #[]_[] #Cavs ...  2019-03-07 01:45:06   
4  🖊️Official Twitter handle of Department of Inf...  2017-02-12 06:45:15   

   user_followers  user_friends  user_favourites  user_verified  \
0             624           950            18775          False   
1            2253          1677               24           True   
2            9275          9525             7254          False   
3             197           987             1488          False   
4          101009           168              101          False   

                  date                                               text  \
0  2020-07-25 12:27:21  If I smelled the scent of hand sanitizers toda...   
1  2020-07-25 12:27:17  Hey @Yankees @YankeesPR and @MLB - wouldn't it...   
2  2020-07-25 12:27:14  @diane3443 @wdunlap @realDonaldTrump Trump nev...   
3  2020-07-25 12:27:10  @brookbanktv The one gift #COVID19 has give me...   
4  2020-07-25 12:27:08  25 July : Media Bulletin on Novel #CoronaVirus...   

                            hashtags               source  is_retweet  
0                                NaN   Twitter for iPhone       False  
1                                NaN  Twitter for Android       False  
2                        ['COVID19']  Twitter for Android       False  
3                        ['COVID19']   Twitter for iPhone       False  
4  ['CoronaVirusUpdates', 'COVID19']  Twitter for Android       False

In [115]:
df_covid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179108 entries, 0 to 179107
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   user_name         179108 non-null  object
 1   user_location     142337 non-null  object
 2   user_description  168822 non-null  object
 3   user_created      179108 non-null  object
 4   user_followers    179108 non-null  int64 
 5   user_friends      179108 non-null  int64 
 6   user_favourites   179108 non-null  int64 
 7   user_verified     179108 non-null  bool  
 8   date              179108 non-null  object
 9   text              179108 non-null  object
 10  hashtags          127774 non-null  object
 11  source            179031 non-null  object
 12  is_retweet        179108 non-null  bool  
dtypes: bool(2), int64(3), object(8)
memory usage: 15.4+ MB


Для проверки актуальности твита нас интересует только дата, поэтому в столбце date отрежем время публикации.

In [120]:
df_covid['date'] = [datetime.strptime(a, "%Y-%m-%d %H:%M:%S").date() for a in df_covid['date']]

Твиты будем считать актуальными, если с момента их публикации до настоящего момента прошло менее 14 дней. 

In [122]:
def is_fresh(df):
    today = date.today()
    list = []
    
    for a in df['date']:
        if (today - a).days > 14:
            list.append('Outdated')
        else:
            list.append('Fresh')
    df['Freshness'] = list

is_fresh(df_covid)

df_covid['Freshness'].value_counts()

Freshness
Outdated    179108
Name: count, dtype: int64

Отсюда видим, что все посты были сделаны больше 14 дней назад. Заменим столбец дат на случайно сгенерированные даты, начиная с 25.07.2020 по сегодняшний день.

In [124]:
start_date = df_covid['date'][0]
end_date = date.today()
num_dates = df_covid['date'].count()

random_dates = [start_date + timedelta(days=random.randint(0, (end_date - start_date).days)) for _ in range(num_dates)]

df_covid['date'] = random_dates

Посмотрим на новый датасет

In [47]:
df_covid.head()

user_name         user_location  \
0           ᏉᎥ☻լꂅϮ            astroworld   
1    Tom Basile 🇺🇸          New York, NY   
2  Time4fisticuffs      Pewee Valley, KY   
3      ethel mertz  Stuck in the Middle    
4         DIPR-J&K     Jammu and Kashmir   

                                    user_description         user_created  \
0  wednesday addams as a disney princess keepin i...  2017-05-26 05:46:42   
1  Husband, Father, Columnist & Commentator. Auth...  2009-04-16 20:06:23   
2  #Christian #Catholic #Conservative #Reagan #Re...  2009-02-28 18:57:41   
3  #Browns #Indians #ClevelandProud #[]_[] #Cavs ...  2019-03-07 01:45:06   
4  🖊️Official Twitter handle of Department of Inf...  2017-02-12 06:45:15   

   user_followers  user_friends  user_favourites  user_verified        date  \
0             624           950            18775          False  2024-03-03   
1            2253          1677               24           True  2021-10-05   
2            9275          9525             7254          False  2023-10-11   
3             197           987             1488          False  2022-07-20   
4          101009           168              101          False  2022-08-26   

                                                text  \
0  If I smelled the scent of hand sanitizers toda...   
1  Hey @Yankees @YankeesPR and @MLB - wouldn't it...   
2  @diane3443 @wdunlap @realDonaldTrump Trump nev...   
3  @brookbanktv The one gift #COVID19 has give me...   
4  25 July : Media Bulletin on Novel #CoronaVirus...   

                            hashtags               source  is_retweet  \
0                                NaN   Twitter for iPhone       False   
1                                NaN  Twitter for Android       False   
2                        ['COVID19']  Twitter for Android       False   
3                        ['COVID19']   Twitter for iPhone       False   
4  ['CoronaVirusUpdates', 'COVID19']  Twitter for Android       False   

  Freshness  
0  Outdated  
1  Outdated  
2  Outdated  
3  Outdated  
4  Outdated

И проверим новые данные на актуальность.

In [128]:
is_fresh(df_covid)

df_covid['Freshness'].value_counts()

Freshness
Outdated    177347
Fresh         1761
Name: count, dtype: int64

Теперь у нас есть актуальные твиты.

In [53]:
df_covid[df_covid['Freshness']=='Fresh']

user_name          user_location  \
107          Dr. Miguel Escalon           New York, NY   
276            Michelle Dunning  Wales, United Kingdom   
332         Coronavirus Updates                    NaN   
559     Tired Healthcare Worker                    NaN   
560                      Robert    Melbourne, Victoria   
...                         ...                    ...   
178625            K[Ash] 🇨🇦🇺🇸🇮🇱                    NaN   
178700                30Seconds                    NaN   
178799          House of Comedy                    NaN   
178929   Unpara//eled Dimension                    NaN   
179103   AJIMATI AbdulRahman O.        Ilorin, Nigeria   

                                         user_description  \
107     Miguel Escalon, MD, MPH: Vice Chair, Director ...   
276     Primary Care Service Manager working in North ...   
332     COVID-19 (Coronavirus) Latest News & Statistic...   
559     Book reader, gardener, Yogi, and activist! #Fr...   
560     PhD Student at Swinburne University of Technol...   
...                                                   ...   
178625  #MCGA #MAGA #KAG #AmYisraelChai #BREXIT #CPC #...   
178700  Inspire & be inspired through the power of con...   
178799                                                NaN   
178929                                                NaN   
179103    Animal Scientist|| Muslim|| Real Madrid/Chelsea   

               user_created  user_followers  user_friends  user_favourites  \
107     2017-10-16 19:50:59            1299           994             2743   
276     2018-01-09 13:02:49             543           970            16055   
332     2020-03-07 12:52:51             433             1                0   
559     2009-10-14 13:33:45            1949          2546            23669   
560     2013-08-11 01:13:12             133           290             1002   
...                     ...             ...           ...              ...   
178625  2011-11-05 14:49:40           16190         17626           122520   
178700  2010-10-22 20:24:14           29488          6316            60231   
178799  2009-06-11 15:47:11            3331           532             3615   
178929  2018-06-30 23:53:14               1             2                3   
179103  2013-12-30 18:59:19             412          1609             1062   

        user_verified        date  \
107             False  2024-10-07   
276             False  2024-10-09   
332             False  2024-10-09   
559             False  2024-09-27   
560             False  2024-10-08   
...               ...         ...   
178625          False  2024-09-29   
178700          False  2024-09-30   
178799          False  2024-10-01   
178929          False  2024-10-01   
179103          False  2024-10-02   

                                                     text  \
107     #coronavirus #COVID19 disproportionately affec...   
276     @BenJamesPhotos A new beginning 6yrs ago.  A w...   
332     1,005 new cases and 18 new deaths in the Unite...   
559     @realDonaldTrump Could you perhaps worry more ...   
560     Making my first interactive dashboard on #Orac...   
...                                                   ...   
178625  🇨🇦 Update:\n\n#Psyop #Propaganda #COVID19 #Sto...   
178700  2020. Saying it's been an unprecedented year i...   
178799  Thank you #FrontlineWorkers 🩺🚑🚒😷for all that y...   
178929  We finally broke free on this Road Trip to Pai...   
179103  Thanks @IamOhmai for nominating me for the @WH...   

                                                 hashtags  \
107      ['coronavirus', 'COVID19', 'Hispanic', 'Latinx']   
276                                                   NaN   
332     ['coronavirus', 'CoronaVirusUpdate', 'COVID19'...   
559                         ['globalpandemic', 'covid19']   
560                                       ['OracleCloud']   
...                                                   ...   
178625  ['Psyop', 'Propaganda', 'COVID19', 'StopTh

## Задание 4. Проверка полноты данных. 

Рассмотрите набор данных о заболеваемости раком, проанализируйте, приведены ли данные по всем странам или некоторые из них отсутствуют. 

Набор данных для использования: "Global Cancer Statistics 2018 (GLOBOCAN 2018)" на Kaggle.
Ссылка: https://www.kaggle.com/ashkhagan/global-cancer-statistics-2018-globocan-2018


In [60]:
df_cancer = pd.read_csv(r'Datasets/cancer.csv')

In [62]:
df_cancer.head()

VAR                      Variable      UNIT  \
0  CANCTOCA           Malignant neoplasms  PERCMTTX   
1  CANCTOCA           Malignant neoplasms  PERCMTTX   
2  CANCTOCA           Malignant neoplasms  PERCMTTX   
3  CANCCOLC  Malignant neoplasms of colon  PERCMTTX   
4  CANCCOLC  Malignant neoplasms of colon  PERCMTTX   

                            Measure  COU    Country   YEA  Year  Value  \
0  Incidence per 100 000 population  AUS  Australia  2002  2002  312.0   
1  Incidence per 100 000 population  AUS  Australia  2008  2008  314.1   
2  Incidence per 100 000 population  AUS  Australia  2012  2012  323.0   
3  Incidence per 100 000 population  AUS  Australia  2002  2002   41.7   
4  Incidence per 100 000 population  AUS  Australia  2008  2008   38.7   

   Flag Codes  Flags  
0         NaN    NaN  
1         NaN    NaN  
2         NaN    NaN  
3         NaN    NaN  
4         NaN    NaN

Введем в рассмотрение новый датасет, состоящий только из стран, удалив дубликаты.

In [65]:
df_countries = pd.DataFrame(data = df_cancer['Country'].unique(), columns=['Country'])

df_countries.count()

Country    44
dtype: int64

Поскольку в данном датасете названия некоторых стран указано неправильно, то проведем некоторую предобработку.

In [68]:
df_countries['Country'] = df_countries['Country'].replace({'China (People\'s Republic of)': 'China', 'Slovak Republic': 'Slovakia', 'Czech Republic':'Czechia', 'Korea':'South Korea'})

Всего у нас собрана информация про 44 страны. Очевидно, это не все страны, для того, чтобы определить, каких стран не хватает, подгрузим список всех стран.

In [71]:
from country_list import countries_for_language

countries = dict(countries_for_language('en'))

df_all_countries = pd.DataFrame(data=countries.values(), columns=['Country'])

df_all_countries.count()

Country    249
dtype: int64

In [73]:
df_countries = pd.concat([df_countries, df_all_countries], ignore_index=True)

df_countries

Country
0          Australia
1            Austria
2            Belgium
3             Canada
4            Czechia
..               ...
288  Wallis & Futuna
289   Western Sahara
290            Yemen
291           Zambia
292         Zimbabwe

[293 rows x 1 columns]

In [75]:
unique_countries = df_countries[~df_countries.duplicated('Country', keep=False)].reset_index()

print(unique_countries)

     index          Country
0       44      Afghanistan
1       45    Åland Islands
2       46          Albania
3       47          Algeria
4       48   American Samoa
..     ...              ...
200    288  Wallis & Futuna
201    289   Western Sahara
202    290            Yemen
203    291           Zambia
204    292         Zimbabwe

[205 rows x 2 columns]


Здесь видно, что в нашем датасете отсутствует инфорамция о 205 странах.